WalletWasabi/WabiSabi/Client/Batching/Payment.cs

```diff
@@ -1,10 +1,35 @@
+using System.Collections.Immutable;
 using WalletWasabi.Extensions;
 using NBitcoin;
+using WalletWasabi.WabiSabi.Models;
 
 namespace WalletWasabi.WabiSabi.Client.Batching;
 
-public abstract record Payment(IDestination Destination, Money Amount)
+public abstract record PaymentState
 {
+	protected PaymentState(PaymentState previousState)
+	{
+		PreviousState = previousState;
+	}
+
+	public PaymentState? PreviousState { get; init; }
+}
+
+public record PendingPayment(PaymentState? PreviousState) : PaymentState(PreviousState);
+public record InProgressPayment(PaymentState PreviousState, uint256 RoundId) : PaymentState(PreviousState);
+public record FinishedPayment(PaymentState PreviousState, uint256 TransactionId) : PaymentState(PreviousState);
+
+public record Payment(IDestination Destination, Money Amount)
+{
+	public Guid Id { get; } = Guid.NewGuid();
+	public PaymentState State { get; init; } = new PendingPayment(null);
```

Comment:
We could avoid using needless null as parameter via using parameterless constructor.

Reply:
We can create a new constructor and move the null there ofc.